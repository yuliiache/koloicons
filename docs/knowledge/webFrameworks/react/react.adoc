== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







