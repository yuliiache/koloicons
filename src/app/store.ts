import {getReducers} from '@kolo/core/reducers/getReducers';
import {rootSagas} from '@kolo/core/sagas/rootSagas';
import {configureStore} from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';

const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
  reducer: getReducers(),
  devTools: process.env.NODE_ENV !== 'production',
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(sagaMiddleware),
});

sagaMiddleware.run(rootSagas);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
