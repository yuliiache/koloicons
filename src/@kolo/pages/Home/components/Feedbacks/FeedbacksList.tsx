import {Feedback} from '@kolo/facade/getFeedbacks/types';
import {FC} from 'react';

import FeedbackItem from './components/FeedbackItem/FeedbackItem';
import styles from './FeedbacksList.module.scss';
import useFeedbacks from './useFeedbacks';

const FeedbacksList: FC = () => {
  const {feedbacks} = useFeedbacks();

  return (
    <section className={styles.feedbacks}>
      <div className={styles.feedbackContainer}>
        {feedbacks &&
          feedbacks.map((feedback: Feedback) => (
            <FeedbackItem
              key={feedback.name}
              name={feedback.name}
              text={feedback.text}
              company={feedback.company}
              avatarPath={feedback.avatarPath}
              url={feedback.url}
            />
          ))}
      </div>
    </section>
  );
};

export default FeedbacksList;
