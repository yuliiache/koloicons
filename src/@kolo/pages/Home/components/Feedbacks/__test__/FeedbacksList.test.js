import React from 'react';
import {Provider} from 'react-redux';
import {useSelector} from 'react-redux';
import {render} from '@testing-library/react';
import store from 'app/store';

import FeedbacksList from '../FeedbacksList';

const mockFeedbacks = [
  {
    name: 'John Doe',
    text: 'This is a sample feedback text.',
    company: 'Sample Company',
    avatarPath: 'avatar.jpg',
    url: 'https://example.com',
  },
  {
    name: 'Jane Smith',
    text: 'Another feedback text.',
    company: 'Another Company',
    avatarPath: 'avatar2.jpg',
    url: 'https://example.com',
  },
];

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));

describe('FeedbacksList Component', () => {
  it('renders snapshot correctly', () => {
    useSelector.mockReturnValue(mockFeedbacks);

    const {asFragment} = render(
      <Provider store={store}>
        <FeedbacksList />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
