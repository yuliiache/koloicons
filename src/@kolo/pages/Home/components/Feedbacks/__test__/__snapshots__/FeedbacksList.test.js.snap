// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbacksList Component renders snapshot correctly 1`] = `
<DocumentFragment>
  <section
    class="feedbacks"
  >
    <div
      class="feedbackContainer"
    >
      <div
        class="feedbackItem"
      >
        <img
          alt="John Doe"
          height="56"
          src="undefined/avatar.jpg"
          width="56"
        />
        <p
          class="feedbackText"
          data-testid="feedbackText"
        >
          This is a sample feedback text.
        </p>
        <span>
          <a
            class="link link--default"
            data-testid="externalLinkComponent"
            href="https://example.com"
            rel="noopener noreferrer nofollow"
          >
            John Doe, Sample Company
          </a>
        </span>
      </div>
      <div
        class="feedbackItem"
      >
        <img
          alt="Jane Smith"
          height="56"
          src="undefined/avatar2.jpg"
          width="56"
        />
        <p
          class="feedbackText"
          data-testid="feedbackText"
        >
          Another feedback text.
        </p>
        <span>
          <a
            class="link link--default"
            data-testid="externalLinkComponent"
            href="https://example.com"
            rel="noopener noreferrer nofollow"
          >
            Jane Smith, Another Company
          </a>
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
